lib/ngx/re.lua

```diff
@@ -0,0 +1,169 @@
+-- I hereby assign copyright in this code to the lua-resty-core project,
+-- to be licensed under the same terms as the rest of the code.
+
+
+local ffi = require 'ffi'
+local bit = require "bit"
+local base = require "resty.core.base"
+local core_regex = require "resty.core.regex"
+
+
+local C = ffi.C
+local sub = string.sub
+local type = type
+local band = bit.band
+local new_tab = base.new_tab
+local tostring = tostring
+local math_max = math.max
+local math_min = math.min
+local math_huge = math.huge
+local re_match_compile = core_regex.re_match_compile
+local destroy_compiled_regex = core_regex.destroy_compiled_regex
+
+
+local FLAG_DFA               = 0x02
+local PCRE_ERROR_NOMATCH     = -1
+local DEFAULT_SPLIT_RES_SIZE = 4
+
+
+local split_ctx = new_tab(0, 1)
+
+
+local _M = { version = base.version }
+
+
+local function re_split_helper(subj, compiled, compile_once, flags, ctx)
+    local rc
+    do
+        local pos = math_max(ctx.pos - 1, 0)
+
+        rc = C.ngx_http_lua_ffi_exec_regex(compiled, flags, subj, #subj, pos)
+    end
+
+    if rc == PCRE_ERROR_NOMATCH then
+        if not compile_once then
+            destroy_compiled_regex(compiled)
+        end
+        return nil, nil, nil
+    end
+
+    if rc < 0 then
+        if not compile_once then
+            destroy_compiled_regex(compiled)
+        end
+        return nil, nil, nil, "pcre_exec() failed: " .. rc
+    end
+
+    if rc == 0 then
+        if band(flags, FLAG_DFA) == 0 then
+            return nil, nil, nil, "capture size too small"
+        end
+
+        rc = 1
+    end
+
+    local caps = compiled.captures
+    local ncaps = compiled.ncaptures
+
+    local from = caps[0] + 1
+    local to = caps[1]
+
+    if from < 0 or to < 0 then
+        return nil, nil, nil
+    end
+
+    ctx.pos = to + 1
+
+    -- retrieve the first sub-match capture if any
+
+    if ncaps > 0 and rc > 1 then
+        return from, to, sub(subj, caps[2] + 1, caps[3])
+    end
+
+    return from, to
+end
+
+
+function _M.split(subj, regex, opts, ctx, max, res)
+    -- we need to cast this to strings to avoid exceptions when they are
+    -- something else.
+    -- needed because of further calls to string.sub in this function.
+    subj = tostring(subj)
+
+    if not ctx then
+        ctx = split_ctx
+        ctx.pos = 1 -- set or reset upvalue field
+
+    elseif not ctx.pos then
+        -- ctx provided by user but missing pos field
+        ctx.pos = 1
+    end
+
+    if not max or max <= 0 then
+        -- mimic perl's split behavior where max <= 0
+        -- translates to "unlimited"
+        max = math_huge
```

Comment:
I don't like the `math.huge` thing since in dual-number mode, the value in `max` may well be a 32-bit integer instead of a double. Mixing it with `math.huge` may defeat such opportunities of optimizations.